sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,755 @@
+
+[source, c++]
+----
+void set_new_threshold(size_t newThreshold);
+----
+
+_Effects_: Sets the deallocation threshold of the memory pool if the value of
+`newThreshold` is larger than the current threshold.
+
+_Throws_: An exception with the `errc::invalid` error code if the value of
+`newThreshold` is lower than the current threshold or larger than the maximum
+allocation size.
+
+
+=== Memory pool properties
+
+A memory pool can be constructed with a number of properties which can change
+certain behaviors, these can be specified when constructing a `memory_pool`
+object.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+struct initial_threshold_key {
+  using value_t = property_value<initial_threshold_key>;
+
+  initial_threshold_key(size_t initialThreshold);
+};
+
+struct maximum_size_key {
+  using value_t = property_value<maximum_size_key>;
+
+  initial_threshold_key(size_t maxSize);
+};
+
+struct read_only_key {
+  using value_t = property_value<read_only_key>;
+
+  read_only_key(bool readOnly);
+};
+
+struct zero_init_key {
+  using value_t = property_value<zero_init_key>;
+
+  read_only_key(bool zeroInit);
+};
+
+struct use_existing_memory_key {
+  using value_t = property_value<use_existing_memory_key>;
+
+  use_existing_memory(void *ptr, size_t size);
+};
+
+inline constexpr initial_threshold_key::value_t initial_threshold;
+inline constexpr maximum_size_key::value_t maximum_size;
+inline constexpr read_only_key::value_t read_only;
+inline constexpr zero_init_key::value_t zero_init;
+inline constexpr use_existing_memory_key::value_t use_existing_memory;
+
+}  // ext::oneapi::experimental
+----
+
+|===
+|Property|Description
+
+|`initial_threshold`
+|The `initial_threshold` property specifies the initial deallocation threshold
+ value for the memory pool. If this property is not used the default value is
+ zero, and this can be increased after the memory pool is created by calling
+ `memory_pool::set_new_threshold`.
+
+|`maximum_size`
+|The `maximum_size` property specifies the maximum size of the memory pool,
+ after which any allocation will result in an exception. If the value specified
+ is larger than the implementation can support an exception with the
+ `errc::memory_allocation` error code is thrown. If this property is not used
+ the default value is implementation-defined.
+
+|`read_only`
+|The `read_only` property is a performance hint which asserts that all memory
+ allocations from the memory pool will only ever be read from, this can be used
+ by the SYCL runtime to optimize for performance.
+
+|`zero_init`
+|The `zero_init` property adds the requirement that all memory allocated to the
+ memory as it is allocated from the source to the memory pool will be
+ initialised to zero. Note there is no guarantee that the memory allocation be
+ re-initialized to zero when it is re-allocated from the pool, so users must
+ re-initialize memory to zero if they wish for later allocations to have this
+ behavior.
```

Comment:
I wonder if we really need this property.  It seems like applications can define this semantic themselves by using `use_existing_memory` and zero-initializing the content of the buffer.

Reply:
That's true, you can achieve the zero-init semantics by using the `use_existing_memory` property and zero-initialising the data before passing the pointer to the memory pool, however, using the `use_existing_memory` also means the memory pool looses the ability to dynamically scale the amount of memory allocated as required, which is another important requirement.

Reply:
OK.  I can see that.